zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zeppelin.notebook.repo;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.zeppelin.conf.ZeppelinConfiguration;
+import org.apache.zeppelin.conf.ZeppelinConfiguration.ConfVars;
+import org.apache.zeppelin.notebook.Note;
+import org.apache.zeppelin.notebook.NoteInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Notebook repository sync with remote storage
+ */
+public class NotebookRepoSync implements NotebookRepo{
+  private VFSNotebookRepo localRepo;
+  private NotebookRepo remoteRepo;
+  private static final Logger LOG = LoggerFactory.getLogger(NotebookRepoSync.class);
+
+  /**
+   * @param (conf)
+   * @throws - Exception
+   */
+  public NotebookRepoSync(ZeppelinConfiguration conf) throws Exception {
+    localRepo = new VFSNotebookRepo(conf);
+    final String defStorageClassName = "org.apache.zeppelin.notebook.repo.VFSNotebookRepo";
+    final String nbStorageClassName = conf.getString(ConfVars.ZEPPELIN_NOTEBOOK_STORAGE).trim();
+    
+    if (nbStorageClassName.isEmpty() || nbStorageClassName.equals(defStorageClassName)) {
+      remoteRepo = null;
+    } else {
+      Class<?> notebookStorageClass = getClass().forName(
+                conf.getString(ConfVars.ZEPPELIN_NOTEBOOK_STORAGE));
+      Constructor<?> constructor = notebookStorageClass.getConstructor(
+                ZeppelinConfiguration.class);
+      remoteRepo = (NotebookRepo) constructor.newInstance(conf);
+      localToRemoteSync();
+    }
+
+  }
+
+  public List<NoteInfo> list() throws IOException {
+    return localRepo.list();
+  }
+
+  public Note get(String noteId) throws IOException {
+    return localRepo.get(noteId);
+  }
+
+  public void save(Note note) throws IOException {
+    localRepo.save(note);
+    if (remoteRepo != null) remoteRepo.save(note);
```

Comment:
Using `null` to differentiate between two 'modes of operations' for NotebookRepoSync is not very readable approach - a reader needs to keep track of the meaning of `null` for this variable to understand the flow.

Why do not keep the same invariant but do something like (from the top of my head)
```java
if (remoteRepoIsDefined) {
  remoteRepo.save()
}
```

even better would be

```java
if (remoteRepo.isDefined()) {
  remoteRepo.save()
}
```

BTW, [styleguide/javaguide#s4.1.1-braces-always-used](http://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used)